### PR TITLE
ENH: grouping

### LINKED ISSUE #280: conflicting study identifiers in ABCD data
Hi, I'm using heudiconv to convert some of the ABCD data into BIDS format, but it seems to only be working on data collected on the Siemens scanner. For the other scanners, I'm getting the error below. 

INFO: Running heudiconv version 0.5.2-dev
INFO: Need to process 1 study sessions
INFO: PROCESSING STARTS: {'subject': 'NDARINVE367LYLC', 'outdir': '/output/', 'session': None}
INFO: Processing 49972 dicoms
INFO: Analyzing 49972 dicoms
Traceback (most recent call last):
  File "/opt/miniconda-latest/bin/heudiconv", line 11, in <module>
    load_entry_point('heudiconv', 'console_scripts', 'heudiconv')()
  File "/src/heudiconv/heudiconv/cli/run.py", line 125, in main
    process_args(args)
  File "/src/heudiconv/heudiconv/cli/run.py", line 338, in process_args
    overwrite=args.overwrite,)
  File "/src/heudiconv/heudiconv/convert.py", line 159, in prep_conversion
    grouping=None)
  File "/src/heudiconv/heudiconv/dicoms.py", line 88, in group_dicoms_into_seqinfos
    studyUID, file_studyUID
AssertionError: Conflicting study identifiers found [1.2.840.113619.6.374.258119085267376010175966987442804122141, 1.2.840.113619.6.374.54018679043989047307905145274801870417].

I've tried running dcm2niix on these dicoms outside of heudiconv, and that seems to work fine. 

Thanks for any advice!
David

### LINKED ISSUE #312: rationale behind seqinfo structure
I am trying to use heudiconv to convert all the data from one scanner to bids.
I like the fact that we have access to the full power of python to write complex heuristics but the seqinfo given to infotodict is missing a lot of information from the dicoms that would really help identify runs without relying on fields that can be modified at the console like SeriesDescription or else.

For instance, I would rather get the phase encoding direction from a specific field than from the string '_AP' being included in the SeriesDescription.
The same problem occurs for the infotoids.

Is there a reason to make the seqinfo that sparse?
Are there any plans to provide extension?
What are the 'unspecified2' 'unspecified3' field for?

Thanks!

### LINKED ISSUE #355: Confusing/misleading message when trying to override subject id
```shell
$ heudiconv -f XXX_reproin.py -s sid000861 --ses 1 --files  .../A002222  # session=None, subject='sid000861')
INFO: Running heudiconv version 0.5.4
INFO: Analyzing 3924 dicoms
INFO: Filtering out 0 dicoms based on their filename
/usr/lib/python2.7/dist-packages/heudiconv/dicoms.py:58: UserWarning: The DICOM readers are highly experimental, unstable, and only work for Siemens time-series at the moment
Please use with caution.  We would be grateful for your help in improving them
  import nibabel.nicom.dicomwrappers as dw
INFO: Generated sequence info for 2 studies with 23 entries total
Traceback (most recent call last):
  File "/usr/bin/heudiconv", line 11, in <module>
    load_entry_point('heudiconv==0.5.4', 'console_scripts', 'heudiconv')()
  File "/usr/lib/python2.7/dist-packages/heudiconv/cli/run.py", line 127, in main
    process_args(args)
  File "/usr/lib/python2.7/dist-packages/heudiconv/cli/run.py", line 260, in process_args
    args.subjs, grouping=args.grouping)
  File "/usr/lib/python2.7/dist-packages/heudiconv/parser.py", line 176, in get_study_sessions
    "found %d sequences" % (sids, len(sids), len(seqinfo_dict))
RuntimeError: We were provided some subjects (['sid000861']) but we can deal only with overriding only 1 subject id. Got 1 subjects and found 2 sequences
```
so is it a matter of overriding only in a single sequence? since it is only a single subject id.  So some logic might need to get fixed or extended

### PR DESCRIPTION
**Pull request by mgxd**
Closes #200
Closes #280 
Closes #312 
Closes #355 
maybe others?

This branch expands the capabilities of `--grouping` to include 2 new options: **all** and **custom**
- **all** assigns every DICOM found to the same subject, even if the study UID is different
- **custom** allows for 2 heuristic defined options:
  - a pydicom attribute to groupby assigned as a global variable within the heuristic
  - a method for more advanced groupings and seqinfo manipulations, such as adding additional fields. 

##### heuristic.py
--------------
```python

grouping = 'AcquisitionDate'
...
or
...
def grouping(files, dcmfilter, seqinfo):
    seqinfos = collections.OrderedDict()
    ...
    return seqinfos  # ordered dict containing seqinfo objects: list of DICOMs
```

Some other changes were hit along the way:
- `group_dicoms_into_seqinfos` returned different seqinfo formats depending on `-d / --files` usage. A new keyword arg `flatten` was added to highlight this distinction.
- ignore nibabel import warnings
- ~~`SeqInfo` changes:~~
   - ~~replace `total_files_till_now` field with `series_files`~~
   - ~~remove `unspecified` fields~~
- `get_typed_attr` utility function added which acts as a strongly typed getattr.
- Increased DICOM attribute extraction speed by stopping read before pixel data.

TODO:
- [x] add regression tests

### CODE REVIEW

File: heudiconv/utils.py [hunk 1]
```diff
@@ -21,12 +21,10 @@
 
 
 seqinfo_fields = [
-    'total_files_till_now',  # 0
+    'series_files',  # 0
     'example_dcm_file',      # 1
     'series_id',             # 2
     'dcm_dir_name',          # 3
-    'unspecified2',          # 4
-    'unspecified3',          # 5
```
Comment (yarikoptic) [inline heudiconv/utils.py]: wouldn't this break any hope for backward compatibility?

Comment (mgxd) [reply heudiconv/utils.py]: yes, there are quite a few phased-out designs we could part with.

though we can add some config/argument to re-enable legacy seqinfo fields
some ideas for the next release can be found at https://github.com/nipy/heudiconv/projects/2

Comment (yarikoptic) [reply heudiconv/utils.py]: I am just afraid that many custom heuristics which access fields via positional indexes would get unnecessarily (i.e. we could avoid it) broken.  I would just keep those afloat for now

**Comment by yarikoptic** [comment]
quick one: thanks @mgxd again for the PR and sorry I forgot to re-review it.  Will do asap.
Meanwhile -- why to keep it in Draft mode, any extensions (besides addressing comments) are intended to come?

File: heudiconv/dicoms.py [hunk 2]
```diff
@@ -6,25 +6,154 @@
...
       Applied to each item of filenames. Should return True if file needs to be
       kept, False otherwise.
     dcmfilter : callable, optional
       If called on dcm_data and returns True, it is used to set series_id
-    grouping : {'studyUID', 'accession_number', None}, optional
-        what to group by: studyUID or accession_number
+    flatten : bool, optional
+      Creates a flattened `seqinfo` with corresponding DICOM files. True when
+      invoked with `dicom_dir_template`.
+    custom_grouping: variable or method, optional
```
Comment (yarikoptic) [inline heudiconv/dicoms.py]: so default is `None`. Not clear to me (either from docstring here or from glancing over the code) what would happen if I specify `grouping='custom'` but do not specify `custom_grouping`.  Please clarify here if there is an intended behavior. If not - add a check that `custom_grouping` is not None or smth alike (e.g. if it is not callable - it is a string)

File: heudiconv/dicoms.py [hunk 3]
```diff
@@ -138,135 +236,72 @@ def group_dicoms_into_seqinfos(files, file_filter, dcmfilter, grouping):
...
         if per_studyUID:
             key = studyUID.split('.')[-1]
-        elif per_accession_number:
-            key = accession_number
+        elif grouping == 'accession_number':
+            key = accession_number = mw.dcm_data.get("AccessionNumber")
+        elif grouping == 'all':
+            key = 'all'
+        elif custom_grouping:
+            key = custom = mw.dcm_data.get(custom_grouping)
```
Comment (yarikoptic) [inline heudiconv/dicoms.py]: it seems that `custom` is used only as a `key`down in the code, i.e. `key` wouldn't change.  So should there be no `custom` and just `key`?

File: heudiconv/tests/test_regression.py [hunk 4]
```diff
@@ -81,3 +82,47 @@ def test_multiecho(tmpdir, subject='MEEPI', heuristic='bids_ME.py'):
...
+    grouping=None)
+../dicoms.py:96: in group_dicoms_into_seqinfos
+    studyUID, file_studyUID
+E   AssertionError: Conflicting study identifiers found [1.3.12.2.1107.5.2.32.35131.30000014022817282751500000052, 1.3.12.2.1107.5.2.43.66112.30000016081515591223100000010].
+"""
+        runner(hargs)
+
+    # group all found DICOMs under subject, despite conflicts
+    hargs += ["-g", "all"]
+    runner(hargs)
```
Comment (yarikoptic) [inline heudiconv/tests/test_regression.py]: may be things are just skipped?  could we check the output data to be all converted as expected?

File: heudiconv/utils.py [hunk 5]
```diff
@@ -21,28 +21,28 @@
...
-    'referring_physician_name', # 17
-    'series_description',    # 18
-    'sequence_name',         # 19
-    'image_type',            # 20
-    'accession_number',      # 21
-    'patient_age',           # 22
-    'patient_sex',           # 23
-    'date',                  # 24
-    'series_uid'             # 25
+    'total_files_till_now',
```
Comment (yarikoptic) [inline heudiconv/utils.py]: I brought indexes in the comments back, see d662718eceb93465b46bfbee5eaf7a5f42840711 for reasoning

**Comment by yarikoptic** [comment]
@leej3 - would you be interested to try on your use cases?  This one is likely to supersede our elderly attempt at it (#200)

**Comment by leej3** [comment]
I wouldn't be free to do anything more than the most basic testing on this, those datasets already ascended to their happy place in openneuro. @agt24 this might be worth keeping in mind for the next dataset you guys set free!

**Comment by agt24** [comment]
yep, we'll definitely take it for a spin sometime soon!

**Comment by yarikoptic** [comment]
Thank you @mgxd! I will give it a look and a try in a day or two.

**Comment by mgxd** [comment]
@yarikoptic if you have the time, would love to see this tested with some of your own data

File: heudiconv/tests/test_regression.py (hunk 4 above)
Comment (yarikoptic) [reply heudiconv/tests/test_regression.py]: this comment was not addressed yet. Let's make tests stronger -- this is a big and important change, so better be tested thoroughly

Comment (mgxd) [reply heudiconv/tests/test_regression.py]: i addressed this in my latest commit 👍

**Comment by yarikoptic** [comment]
Something weird I cannot grasp.

I am trying to do manual regression testing using that helper script we have (pushed some enhancements).
So comparing 0.5.4 and master works fine:
```
$> heudiconv/utils/test-compare-two-versions.sh heudiconv-{0.5.x,master} --bids -f reproin --files dartmouth-phantoms/bids_test4-20161014/phantom-1
Running /home/yoh/picts/mris/heudiconv-0.5.x/venvs/dev3/bin/heudiconv with log in compare-versions/v0.5.4.log
Running /home/yoh/picts/mris/heudiconv-master/venvs/dev3/bin/heudiconv with log in compare-versions/v0.5.4-85-g135ccb9.log
Results differ: /home/yoh/picts/mris/compare-versions/v0.5.4-v0.5.4-85-g135ccb9.diff
 .heudiconv/phantom1sid1/ses-localizer/info/heuristic.py                  |  951 ++++++++++
 .heudiconv/phantom1sid1/ses-localizer/info/reproin.py                    |  924 ---------
 participants.json                                                        |   14 
 sub-phantom1sid1/ses-localizer/sub-phantom1sid1_ses-localizer_scans.json |   16 
 4 files changed, 981 insertions(+), 924 deletions(-)
```
and changes upon inspection are ok.  But when I try to run/compare against this branch, it errors out:
```
(git)lena:~/picts/mris[master]git
$> heudiconv/utils/test-compare-two-versions.sh heudiconv-{master,fix-group} --bids -f reproin --files dartmouth-phantoms/bids_test4-20161014/phantom-1
Not running heudiconv since compare-versions/v0.5.4-85-g135ccb9 already exists
Running /home/yoh/picts/mris/heudiconv-fix-group/venvs/dev3/bin/heudiconv with log in compare-versions/v0.5.4-97-g5c82c6d.log
Exited with 1  Check compare-versions/v0.5.4-97-g5c82c6d.log
1 7472.....................................:Wed 27 Nov 2019 08:57:15 PM EST:.
(git)lena:~/picts/mris[master]git
$> tail -n 20 compare-versions/v0.5.4-97-g5c82c6d.log
INFO: [Node] Finished "convert".
Traceback (most recent call last):
  File "/home/yoh/picts/mris/heudiconv-fix-group/venvs/dev3/bin/heudiconv", line 11, in <module>
    load_entry_point('heudiconv', 'console_scripts', 'heudiconv')()
  File "/home/yoh/picts/mris/heudiconv-fix-group/heudiconv/cli/run.py", line 127, in main
    process_args(args)
  File "/home/yoh/picts/mris/heudiconv-fix-group/heudiconv/cli/run.py", line 345, in process_args
    grouping=args.grouping,)
  File "/home/yoh/picts/mris/heudiconv-fix-group/heudiconv/convert.py", line 208, in prep_conversion
    dcmconfig=dcmconfig,)
  File "/home/yoh/picts/mris/heudiconv-fix-group/heudiconv/convert.py", line 296, in convert
    overwrite=overwrite)
  File "/home/yoh/picts/mris/heudiconv-fix-group/heudiconv/convert.py", line 605, in save_converted_files
    safe_copyfile(fl, outfile, overwrite)
  File "/home/yoh/picts/mris/heudiconv-fix-group/heudiconv/utils.py", line 371, in safe_copyfile
    % (src, dest)
RuntimeError: was asked to copy /home/yoh/.tmp/dcm2niix1g2rtl7g/convert/home/yoh/picts/mris/compare-versions/v0.5.4-97-g5c82c6d/Halchenko/Yarik/950_bids_test4/sub-phantom1sid1/ses-localizer/func/sub-phantom1sid1_ses-localizer_task-retino_acq-mocoXp2X3mm_run-07_bold_e2_ph.nii.gz but destination already exists: /home/yoh/picts/mris/compare-versions/v0.5.4-97-g5c82c6d/Halchenko/Yarik/950_bids_test4/sub-phantom1sid1/ses-localizer/func/sub-phantom1sid1_ses-localizer_task-retino_acq-mocoXp2X3mm_run-07_echo-2_bold.nii.gz
Command exited with non-zero status 1
1.69user 0.19system 0:04.50elapsed 41%CPU (0avgtext+0avgdata 89124maxresident)k
0inputs+4976outputs (0major+132622minor)pagefaults 0swaps
1 7473.....................................:Wed 27 Nov 2019 08:57:26 PM EST:.
(git)lena:~/picts/mris[master]git
$> ls -ld /home/yoh/picts/mris/compare-versions/v0.5.4-97-g5c82c6d/Halchenko/Yarik/950_bids_test4/sub-phantom1sid1/ses-localizer/func/sub-phantom1sid1_ses-localizer_task-retino_acq-mocoXp2X3mm_run-07_echo-2_bold.nii.gz
-rw------- 1 yoh yoh 6614 Nov 27 20:57 /home/yoh/picts/mris/compare-versions/v0.5.4-97-g5c82c6d/Halchenko/Yarik/950_bids_test4/sub-phantom1sid1/ses-localizer/func/sub-phantom1sid1_ses-localizer_task-retino_acq-mocoXp2X3mm_run-07_echo-2_bold.nii.gz
```
but the weird part is that what seems to trigger it is me having dcmdump files in there... If I clean then up, things start to work:
```shell
191127-21:12:22,492 nipype.interface INFO:
	 stdout 2019-11-27T21:12:22.491940:Conversion required 0.002282 seconds (0.002280 for core code).
INFO: [Node] Finished "convert".
191127-21:12:22,510 nipype.workflow INFO:
	 [Node] Finished "convert".
Traceback (most recent call last):
  File "/home/yoh/picts/mris/heudiconv-fix-group/venvs/dev3/bin/heudiconv", line 11, in <module>
    load_entry_point('heudiconv', 'console_scripts', 'heudiconv')()
  File "/home/yoh/picts/mris/heudiconv-fix-group/heudiconv/cli/run.py", line 127, in main
    process_args(args)
  File "/home/yoh/picts/mris/heudiconv-fix-group/heudiconv/cli/run.py", line 345, in process_args
    grouping=args.grouping,)
  File "/home/yoh/picts/mris/heudiconv-fix-group/heudiconv/convert.py", line 208, in prep_conversion
    dcmconfig=dcmconfig,)
  File "/home/yoh/picts/mris/heudiconv-fix-group/heudiconv/convert.py", line 296, in convert
    overwrite=overwrite)
  File "/home/yoh/picts/mris/heudiconv-fix-group/heudiconv/convert.py", line 605, in save_converted_files
    safe_copyfile(fl, outfile, overwrite)
  File "/home/yoh/picts/mris/heudiconv-fix-group/heudiconv/utils.py", line 371, in safe_copyfile
    % (src, dest)
RuntimeError: was asked to copy /home/yoh/.tmp/dcm2niix3mu7ncqx/convert/tmp/out/Halchenko/Yarik/950_bids_test4/sub-phantom1sid1/ses-localizer/func/sub-phantom1sid1_ses-localizer_task-retino_acq-mocoXp2X3mm_run-07_bold_e2_ph.nii.gz but destination already exists: /tmp/out/Halchenko/Yarik/950_bids_test4/sub-phantom1sid1/ses-localizer/func/sub-phantom1sid1_ses-localizer_task-retino_acq-mocoXp2X3mm_run-07_echo-2_bold.nii.gz

$> git -C dartmouth-phantoms/bids_test4-20161014/phantom-1 clean -dfx
Removing MoCoSeries/1.3.12.2.1107.5.2.43.66112.2016101409265217734302222.dcm.dump
Removing MoCoSeries/1.3.12.2.1107.5.2.43.66112.2016101409265385049702228.dcm.dump
Removing MoCoSeries/1.3.12.2.1107.5.2.43.66112.2016101409270575369702242.dcm.dump
Removing MoCoSeries/1.3.12.2.1107.5.2.43.66112.2016101409270823880402248.dcm.dump
Removing MoCoSeries/1.3.12.2.1107.5.2.43.66112.2016101409272069411002262.dcm.dump
Removing MoCoSeries/1.3.12.2.1107.5.2.43.66112.2016101409272317892602268.dcm.dump
Removing MoCoSeries/1.3.12.2.1107.5.2.43.66112.2016101409273584407202282.dcm.dump
Removing MoCoSeries/1.3.12.2.1107.5.2.43.66112.2016101409273832740502288.dcm.dump

$> rm -rf out; mkdir -p out; ~/picts/mris/heudiconv-fix-group/venvs/dev3/bin/heudiconv --bids -f reproin -o $PWD/out --files dartmouth-phantoms/bids_test4-20161014/phantom-1 | tee bad-fixed.log
...
INFO: Converting /tmp/out/Halchenko/Yarik/950_bids_test4/sub-phantom1sid1/ses-localizer/anat/sub-phantom1sid1_ses-localizer_scout__dup-03 (3 DICOMs) -> /tmp/out/Halchenko/Yarik/950_bids_test4/sub-phantom1sid1/ses-localizer/anat . Converter: dcm2niix . Output types: ('dicom',)
INFO: Populating template files under /tmp/out/Halchenko/Yarik/950_bids_test4
INFO: PROCESSING DONE: {'subject': 'phantom1sid1', 'outdir': '/tmp/out/Halchenko/Yarik/950_bids_test4', 'session': 'localizer'}
~/picts/mris/heudiconv-fix-group/venvs/dev3/bin/heudiconv --bids -f reproin -  12.28s user 0.24s system 83% cpu 15.013 total
```
I don't see any 'dump' files being mentioned in the bad.log.

FWIW, I share that tarball (large -- 300MB, didn't bother filtering - http://www.onerussian.com/tmp/dartmouth-phantoms.yarikweird.tgz
 Let's see if reproducible)... actually if I just take `phantom-1` (even dirty) then things work again... so there is something weird.  my wild guess some non-deterministic dict ordering based on something (like traversed files) triggers some incorrect handling.  I would need to get back to reviewing the diff or trying just to grasp what is going on.  But may be @mgxd has an immediate clue on what is going on

PS that nipype logs flood -- even if I tune HEUDICONV_LOG_LEVEL=ERROR env var -- still falls upon me

**Comment by yarikoptic** [comment]
Hi @mgxd .  I have pushed a merge of master (in 6bd6b246c345247a976179c3e3ff8346b299e31d) which had one conflict already, which I hope I resolved correctly.  I wonder if you would get a chance some time soon to address the existing comments, so we could progress on this valuable PR?!

**Comment by mgxd** [comment]
@yarikoptic aha, there was a bug hiding in there! thanks for the test, I was able to replicate and (hopefully) squash it

tests are now failing due to the latest release of dcmstack not keeping up with nibabel (see https://github.com/moloney/dcmstack/issues/66). some options to consider, in order of effort:
1) port over the code we are using https://github.com/nipy/heudiconv/blob/b3977072f36e71ea0af6484f125cbf561d3f0787/heudiconv/dicoms.py#L391 and https://github.com/nipy/heudiconv/blob/b3977072f36e71ea0af6484f125cbf561d3f0787/heudiconv/dicoms.py#L521 and remove the dependency altogether
2) drop dcmstack and depend only on dcm2niix metadata
3) pin to older nibabel until a release is pushed

**Comment by yarikoptic** [comment]
pushed a quick temporary workaround fix to get tests going.
Considering to drop dcmstack is imho worthwhile, but as a separate issue/PR yet to be investigated in greater detail ;)

**Comment by mgxd** [comment]
@yarikoptic tests are working again 😎 ready for another review

**Comment by mgxd** [comment]
@yarikoptic pinging here again - any last words before we get this in?

**Comment by yarikoptic** [comment]
sorry... will do a few regression tests now.  Did you try yourself with `heudiconv/utils/test-compare-two-versions.sh`?

**Comment by yarikoptic** [comment]
ok, I have done testing on a few sample cases... seems to work without unexpected differences.  I guess lets proceed.  Thank you @mgxd !!!!

**Pull request merged**